src/Libraries/Microsoft.Extensions.AI.Evaluation.NLP/Common/GLEUAlgorithm.cs

```diff
@@ -0,0 +1,62 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.Shared.Diagnostics;
+
+namespace Microsoft.Extensions.AI.Evaluation.NLP.Common;
+
+internal static class GLEUAlgorithm
+{
+    internal static double SentenceGLEU(string[][] references, string[] hypothesis, int minN = 1, int maxN = 4)
+    {
+        if (references == null || references.Length == 0)
+        {
+            Throw.ArgumentNullException(nameof(references), $"'{nameof(references)}' cannot be null or empty.");
+        }
+
+        if (hypothesis == null || hypothesis.Length == 0)
+        {
+            Throw.ArgumentNullException(nameof(hypothesis), $"'{nameof(hypothesis)}' cannot be null or empty.");
+        }
+
+        MatchCounter<NGram<string>> hypNGrams = new(hypothesis.AsSpan().CreateAllNGrams(minN, maxN));
+        int truePosFalsePos = hypNGrams.Sum();
+
+        List<(int, int)> hypCounts = [];
+        foreach (var reference in references)
+        {
+            MatchCounter<NGram<string>> refNGrams = new(reference.AsSpan().CreateAllNGrams(minN, maxN));
```

Comment:
This applies to all the Create*NGrams helpers which currently require the caller to include .AsSpan() in the call site. Coud we add overloads for Create*NGrams that are based on arrays / IEnumerable so that the call sites can omit the AsSpan() calls and just say reference.CreateNGrams / reference.CreateAllNGrams?